src/relay/pass/pass_manager.cc

```diff
@@ -0,0 +1,540 @@
+  PassInfo pass_info;
+
+  /*! \brief The pass function sketches the real optimization. For example,
+   * we may need to perform dead code elimination on the module level. We could
+   * implement the algorithm in the `pass_func` and let it run on a module. It
+   * will then remove the dead code including the unused functions in the module.
+   */
+  runtime::TypedPackedFunc<Module(Module, PassContext)> pass_func;
+
+  ModulePassNode() = default;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+    v->Visit("pass_info", &pass_info);
+  }
+
+  /*!
+   * \brief Run a module pass on a certain module.
+   *
+   * \param mod The module that an optimization pass runs on.
+   *
+   * \return Return the updated module.
+   */
+  Module operator()(const Module& mod) const final;
+
+  /*!
+   * \brief Get the pass information/meta data.
+   */
+  PassInfo GetPassInfo() const { return pass_info; }
+
+  /*!
+   * \brief Set the context information for a module pass.
+   *
+   * \param pass_ctx The context information for a module pass.
+   */
+  void SetContext(const PassContext& pass_ctx) final;
+
+  TVM_DLL static ModulePass make(
+      PassInfo pass_info,
+      runtime::TypedPackedFunc<Module(Module, PassContext)> pass_func);
+
+  static constexpr const char* _type_key = "relay.ModulePass";
+  TVM_DECLARE_NODE_TYPE_INFO(ModulePassNode, PassNode);
+
+ private:
+  /*!
+   * \brief The context information that is used to help perform a module pass.
+   */
+  PassContext pass_ctx_;
+};
+
+RELAY_DEFINE_NODE_REF(ModulePass, ModulePassNode, Pass);
+
+class FunctionPass;
+
+/*!
+ * \brief Function-level passes are used to implement various global
+ * optimizations for a given Relay module. It fetches one function at a time
+ * from the function list in the module for optimization.
+ *
+ * Note that the scope of passes at this level is a Relay function. Therefore,
+ * we cannot add or delete a function through these passes as they are not aware
+ * of the global information.
+ */
+class FunctionPassNode : public PassNode {
+ public:
+  /* \brief The pass meta data.*/
+  PassInfo pass_info;
+
+  /*! \brief The packed pass function sketches the real optimization. For
+   * instance, we can implement a pass that works on a Relay function as a
+   * `pass_func` and let it run on a given module. The same `pass_func` will
+   * then be applied on each function in the module.
+   */
+  runtime::TypedPackedFunc<Function(Function, PassContext)> pass_func;
+
+  FunctionPassNode() = default;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+    v->Visit("pass_info", &pass_info);
+  }
+
+  /*!
+   * \brief Run a function pass on a certain module.
+   *
+   * \param mod The module that an optimization pass runs on.
+   *
+   * \return Return the updated module.
+   */
+  Module operator()(const Module& mod) const final;
+
+  /*!
+   * \brief Get the pass information/meta data.
+   */
+  PassInfo GetPassInfo() const { return pass_info; }
```

Comment:
I am debating that perhaps we can directly rename it to 
```c++
PassInfo info() const;
```

So that it is member property style

Reply:
done